systest/tests/smeshing_test.go

```diff
@@ -0,0 +1,120 @@
+package tests
+
+import (
+	"bytes"
+	"sort"
+	"testing"
+
+	spacemeshv1 "github.com/spacemeshos/api/release/go/spacemesh/v1"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spacemeshos/go-spacemesh/systest/cluster"
+	"github.com/spacemeshos/go-spacemesh/systest/testcontext"
+)
+
+func TestSmeshing(t *testing.T) {
+	tctx := testcontext.New(t, testcontext.Labels("sanity"))
+	cl, err := cluster.Default(tctx, cluster.WithKeys(10))
+	require.NoError(t, err)
+
+	t.Run("Proposals", func(t *testing.T) {
+		t.Parallel()
+		testSmeshing(t, tctx, cl)
+	})
+	t.Run("Transactions", func(t *testing.T) {
+		t.Parallel()
+		testTransactions(t, tctx, cl)
+	})
+}
+
+func testSmeshing(t *testing.T, tctx *testcontext.Context, cl *cluster.Cluster) {
+	const limit = 15
+
+	createdch := make(chan *spacemeshv1.Proposal, cl.Total()*limit)
+	includedAll := make([]map[uint32][]*spacemeshv1.Proposal, cl.Total())
+	for i := 0; i < cl.Total(); i++ {
+		includedAll[i] = map[uint32][]*spacemeshv1.Proposal{}
+	}
+
+	eg, ctx := errgroup.WithContext(tctx)
+	for i := 0; i < cl.Total(); i++ {
+		i := i
+		client := cl.Client(i)
+		collectProposals(ctx, eg, cl.Client(i), func(proposal *spacemeshv1.Proposal) (bool, error) {
+			tctx.Log.Debugw("received proposal event",
+				"client", client.Name,
+				"layer", proposal.Layer.Number,
+				"smesher", prettyHex(proposal.Smesher.Id),
+				"eligibilities", len(proposal.Eligibilities),
+				"status", spacemeshv1.Proposal_Status_name[int32(proposal.Status)],
+			)
+			if proposal.Layer.Number > limit {
+				return false, nil
+			}
+			if proposal.Status == spacemeshv1.Proposal_Created {
+				createdch <- proposal
+			} else {
+				includedAll[i][proposal.Layer.Number] = append(includedAll[i][proposal.Layer.Number], proposal)
+			}
+			return true, nil
+		})
+	}
+
+	require.NoError(t, eg.Wait())
+	close(createdch)
+
+	created := map[uint32][]*spacemeshv1.Proposal{}
+	beacons := map[uint64]map[string]struct{}{}
+	for proposal := range createdch {
+		created[proposal.Layer.Number] = append(created[proposal.Layer.Number], proposal)
+		if edata := proposal.GetData(); edata != nil {
+			if _, exist := beacons[proposal.Epoch.Value]; !exist {
+				beacons[proposal.Epoch.Value] = map[string]struct{}{}
+			}
+			beacons[proposal.Epoch.Value][prettyHex(edata.Beacon)] = struct{}{}
+		}
+	}
+	requireEqualEligibilities(t, created)
+	for layer := range created {
```

Comment:
79-97: perhaps create a method (much like requireEqualEligibilities) to explain what this is verifying. i dont fully understand.